src/Namshi/JOSE/JWS.php

```diff
@@ -24,7 +24,17 @@ class JWS extends JWT
      */
     public function __construct($algorithm, $type = null)
     {
-        parent::__construct(array(), array('alg' => $algorithm, 'typ' => $type ?: "JWS"));
+        if (is_string($algorithm)) {
+            $header = array('alg' => $algorithm, 'typ' => $type ?: "JWS");
+        }
+        else {
+            $header = $algorithm;
+            if (! isset($header['typ'])) {
+                $header['typ'] = $type ?: "JWS";
+            }
+        }
+
+        parent::__construct(array(), $header);
```

Comment:
TBH I think this will be confusing, as the algorith can either be the algorithm itself or a dictionary or headers. I think we should change the constructor rather than have this as it leads to misunderstanding and legacy